compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersOps.td

```diff
@@ -0,0 +1,237 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_DIALECT_MODULES_IO_PARAMETERS_OPS
+#define IREE_DIALECT_MODULES_IO_PARAMETERS_OPS
+
+include "iree/compiler/Dialect/HAL/IR/HALBase.td"
+include "iree/compiler/Dialect/Util/IR/UtilAttrs.td"
+include "iree/compiler/Dialect/Util/IR/UtilInterfaces.td"
+include "iree/compiler/Modules/IO/Parameters/IR/IOParametersBase.td"
+include "mlir/IR/OpAsmInterface.td"
+include "mlir/IR/SymbolInterfaces.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class IOParameters_PureOp<string mnemonic, list<Trait> traits = []> :
+    IOParameters_Op<mnemonic, !listconcat(traits, [Pure])>;
+
+//===----------------------------------------------------------------------===//
+// Parameter I/O
+//===----------------------------------------------------------------------===//
+
+def OpGroupParameterOps : OpDocGroup {
+  let summary = "Parameter I/O ops";
+  let description = "Ops parameter I/O.";
+}
+
+let opDocGroup = OpGroupParameterOps in {
+
+def IOParameters_LoadOp : IOParameters_Op<"load", [
+  Util_SizeAwareOp,
+]> {
+  let summary = [{reads a parameter from a parameter scope}];
+  let description = [{
+    Asynchronously reads a parameter from an external parameter provider and
+    returns the resulting buffer. Depending on the parameter and buffer types
+    this may alias existing cached storage or be directly mapped to the
+    parameter origin or result in a copy as if an allocate + read had been used.
+  }];
+
+  let arguments = (ins
+    HAL_Device:$device,
+    HAL_DeviceQueueAffinity:$queue_affinity,
+    HAL_Fence:$wait_fence,
+    HAL_Fence:$signal_fence,
```

Comment:
So nice to see wait/signal fences on all these ops, since that gives us pipelining and synchronization naturally :)

Reply:
yeah! makes it not usable from the inline hal but that's ok - an inline parameters module would be simpler/smaller and could still reuse all the file format parsing stuff (just not the parameter_provider interface that is async-only)